### ISSUE TITLE
False positive diagnostics of a method with the same name provided by unrelated trait

### ISSUE THREAD

**Issue by inflation** [rust-lang/rust-analyzer#12621]
**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)
rust-analyzer version: 0.0.0 (427061da1 2022-06-19)
**rustc version**: (eg. output of `rustc -V`)
1.61.0
**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

While skimming code of the [h3](https://github.com/hyperium/h3) crates, I encountered this annoying diagnostics at `src/client.rs:139`:
<img width="292" alt="image" src="https://user-images.githubusercontent.com/2375962/175245452-c1ff6e0c-8621-4ffc-9a3c-9c35f53815ed.png">
<img width="403" alt="image" src="https://user-images.githubusercontent.com/2375962/175246390-9fec63e9-8312-4d12-8545-cea04d041ea5.png">

Apparently, `rust-analyzer` chooses the `take()` method from `bytes::Buf` trait instead of the `take()` method from `Option`.

**Comment by flodiebold** [comment]
Seems to be some trait solving issue where Chalk thinks everything implements `Buf` because of the `T: quic::OpenStreams<B>` where clause.
```rust
fn foo<T, B>()
where
    T: quic::OpenStreams<B>,
    B: Buf,
{
    let mut o = Some(1);
    o.take();
}
```

**Comment by ChayimFriedman2** [comment]
This is probably fixed by now. If not @inflation please reopen.
